network/proto/impl.go

```diff
@@ -101,31 +127,42 @@ func (p protocol) startAdvertingHashes() {
 	}
 }
 
+func (p protocol) startQueryingPeerDiagnostics() {
```

Comment:
how do these routines stop nicely?

Reply:
When the program dies, the goroutines die with it (not like Java's non-daemon threads). The network engine's `Stop()` is called when the program halts, which in turn shuts down the P2P layer, which closes all active gRPC connections.  Since all access to these connections is protected using mutexes, race conditions should not occur. It could happen that diagnostics are collected but there are no more peer connections to broadcast it to, because the program is halting, but no harm would be done.